### PR TITLE
Allow publishing/unpublishing programming environments

### PR DESCRIPTION
**Pull request by bethanyaconnor**
Step ~1~ 2 of 2 to add the ability to hide and unhide documentation. This feature will allow curriculum writers to add a new programming environment and write the documentation without it being immediately accessible to students. This will be useful as we move Javalab documentation to levelbuilder and will allow us to decouple that process from launching code docs on code studio.

PRs:
1. (#45110) Add a `published` column to `programming_environments`
2. (this PR) Allow editing of `published` and use it to determine access rules for code docs

In particular, this PR:
- Allows curriculum writers to publish/unpublish a programming environment from the programming environment edit page (screenshot below)
- Seeds and serializes `published`
- Hides unpublished programming environments from /programming_environments (which will eventually be /docs)
- Disallows access of any documentation within an unpublished programming environment to non-levelbuilders

New checkbox on programming environment edit page (highlighting is my cursor):
Screen Shot 2022-03-02 at 10 02 02 AM

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/ability.rb [hunk 1]
```diff
@@ -74,8 +76,12 @@ def initialize(user)
...
     end
 
-    can [:show_by_keys], ProgrammingExpression do |expression|
-      can? :read, expression
+    can [:read], ProgrammingEnvironment do |environment|
+      environment.published || user.permission?(UserPermission::LEVELBUILDER)
+    end
+
+    can [:read, :show_by_keys], ProgrammingExpression do |expression|
+      can? :read, expression.programming_environment
```
Comment (jamescodeorg) [inline dashboard/app/models/ability.rb:85]: This logic is fairly straightforward but do we want to add some unit tests to ability_test.rb?

File: dashboard/app/controllers/programming_environments_controller.rb [hunk 2]
```diff
@@ -58,6 +58,7 @@ def update
   def show
     @programming_environment = ProgrammingEnvironment.find_by_name(params[:name])
     return render :not_found unless @programming_environment
+    return head :forbidden unless can?(:read, @programming_environment)
```
Comment (jamescodeorg) [inline dashboard/app/controllers/programming_environments_controller.rb:60]: For my education, how does this relate to `load_and_authorize_resource` at the top of this class?  Does the top-level call check any abilities?

**Review by jamescodeorg — APPROVED** [review]
A few questions, but generally looks good to me!

File: dashboard/app/controllers/programming_environments_controller.rb (hunk 2 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_environments_controller.rb:60]: Oof this sent me down a rabbit hole.

The short answer to your question is (I think) no. However, according to the [CanCanCan docs](https://github.com/CanCanCommunity/cancancan/blob/develop/docs/changing_defaults.md#override-loading), I should be able to define my own method to find `@programming_environment`. Which just doesn't seem to work 🙃 Not sure if it's a version thing or what but it's not hitting the cutom logic without this explicit `can?` call.

I don't understand how the edit, update, etc permissions are working though. I know @dmcavoy has been looking in this area a lot recently so maybe she has an idea?

File: dashboard/app/models/ability.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/app/models/ability.rb:85]: Oh good call! The `ProgrammingEnvironment` logic is mostly covered by new tests in `programming_environments_controller_test` but I'll add a more comprehensive test

File: dashboard/app/controllers/programming_environments_controller.rb (hunk 2 above)
Comment (dmcavoy) [reply dashboard/app/controllers/programming_environments_controller.rb:60]: CanCanCan tries to be smart and figure out the `@programming_environment` on its own but since you want it to find by name instead of id it probably isn't finding the programming environment as part of the load.  I think you might be able to fix this with `  load_and_authorize_resource find_by: :name`. If not you could do something similar to how we set up script where we `set_unit` and then `authorize_resource`

Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_environments_controller.rb:60]: Hmmm I tried that and it didn't seem to be respecting my custom logic in `ability.rb`. Let me try again today with fresh eyes though

Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_environments_controller.rb:60]: Okay, coming back to this. Writing a custom `set_programming_environment` method does work but it still won't respect the logic from ability.rb. When I look at [script](https://github.com/code-dot-org/code-dot-org/blob/599119f5d32edefee9dee177b98d3d2043407e4f/dashboard/app/controllers/scripts_controller.rb#L266) and [lesson](https://github.com/code-dot-org/code-dot-org/blob/599119f5d32edefee9dee177b98d3d2043407e4f/dashboard/app/controllers/lessons_controller.rb#L28), we do explicitly check that the page `can` be viewed by the user. So I'm inclined to leave this as-is (I'm adding the custom `set_programming_environment` method in another PR already)

Comment (dmcavoy) [reply dashboard/app/controllers/programming_environments_controller.rb:60]: I think it's fine to leave it this way. 

The one you pointed to for script is used because we want to render a custom page when you can't access so its a little different case.

Comment (jamescodeorg) [reply dashboard/app/controllers/programming_environments_controller.rb:60]: Sounds good, thanks for taking a look!

Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_environments_controller.rb:60]: Ahh makes sense about script (and why it's also like that for courses). Thanks for the help Dani!
